<a href="https://colab.research.google.com/github/hmyrcmn/AtfStudiosProject/blob/main/Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Sınıflandırma Nedir?: Sınıflandırma, bir veri kümesini belirli sınıflara ayırma işlemidir. Bu sınıflar genellikle etiketler veya temel gerçeklikleri temsil eder.

İkili ve Çok Sınıflı Sınıflandırma: İkili sınıflandırma, veriyi iki sınıfa ayırırken, çok sınıflı sınıflandırma üç veya daha fazla sınıfa ayırır.

Lojistik Regresyon: Lojistik regresyon, doğrusal bir regresyon gibi görünse de, sınıflandırma problemlerini çözmek için kullanılır. Sigmoid işlevi kullanarak sonuçları 0 ile 1 arasında sınırlar ve ikili sınıflandırma için idealdir.

Sınıflandırma Performans Metrikleri: Model performansını değerlendirmek için kullanılan metrikler şunlardır: doğruluk (accuracy), kesinlik (precision), geri çağırma (recall), ve F1 puanı.

Destek Vektör Makineleri (SVM): SVM, sınıflandırma problemlerini çözmek için kullanılan bir başka denetimli öğrenme modelidir. SVM, farklı sınıfları ayıran bir karar sınırı (hiper düzlem) bulur.

Veri Hazırlığı: Veri kümesinin temizlenmesi ve sayısal hale getirilmesi gerekebilir. Etiketler genellikle sayılara dönüştürülür.

Model Eğitme ve Değerlendirme: Veri kümesi tren ve test setlerine ayrılarak bir sınıflandırma modeli eğitilir ve ardından test verisi üzerinde değerlendirilir.

Sonuçlar: Karışıklık matrisi ve sınıflandırma raporu gibi performans metrikleri kullanılarak modelin başarısı değerlendirilir.

In [1]:
import pandas as pd
dataset = pd.read_csv("breast-cancer.csv")


\Bu kod, pandas kütüphanesini içe aktarıyor ve "breast-cancer.csv" adlı bir CSV dosyasından veri setini okuyor. Veri seti kanser teşhisi için kullanılacak olan özellikleri içeriyor.



In [3]:
dataset.shape

(569, 31)

Bu kod, veri setinin boyutunu (satır ve sütun sayısı) gösterir. Bu veri setinde toplam 569 örnek ve 31 özellik bulunmaktadır.

In [5]:
dataset.head()


diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M     1.097064     -2.073335        1.269934   0.984375   
1         M     1.829821     -0.353632        1.685955   1.908708   
2         M     1.579888      0.456187        1.566503   1.558884   
3         M    -0.768909      0.253732       -0.592687  -0.764464   
4         M     1.750297     -1.151816        1.776573   1.826229   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         1.568466          3.283515        2.652874             2.532475   
1        -0.826962         -0.487072       -0.023846             0.548144   
2         0.942210          1.052926        1.363478             2.037231   
3         3.283553          3.402909        1.915897             1.451707   
4         0.280372          0.539340        1.371011             1.428493   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       2.217515  ...      1.886690      -1.359293         2.303601   
1       0.001392  ...      1.805927      -0.369203         1.535126   
2       0.939685  ...      1.511870      -0.023974         1.347475   
3       2.867383  ...     -0.281464       0.133984        -0.249939   
4      -0.009560  ...      1.298575      -1.466770         1.338539   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    2.001237          1.307686           2.616665         2.109526   
1    1.890489         -0.375612          -0.430444        -0.146749   
2    1.456285          0.527407           1.082932         0.854974   
3   -0.550021          3.394275           3.893397         1.989588   
4    1.220724          0.220556          -0.313395         0.613179   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              2.296076        2.750622                 1.937015  
1              1.087084       -0.243890                 0.281190  
2              1.955000        1.152255                 0.201391  
3              2.175786        6.046041                 4.935010  
4              0.729259       -0.868353                -0.397100  

[5 rows x 31 columns]

In [6]:
dataset.tail()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
564         M     2.110995      0.721473        2.060786   2.343856   
565         M     1.704854      2.085134        1.615931   1.723842   
566         M     0.702284      2.045574        0.672676   0.577953   
567         M     1.838341      2.336457        1.982524   1.735218   
568         B    -1.808401      1.221792       -1.814389  -1.347789   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
564         1.041842          0.219060        1.947285             2.320965   
565         0.102458         -0.017833        0.693043             1.263669   
566        -0.840484         -0.038680        0.046588             0.105777   
567         1.525767          3.272144        3.296944             2.658866   
568        -3.112085         -1.150752       -1.114873            -1.261820   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
564      -0.312589  ...      1.901185       0.117700         1.752563   
565      -0.217664  ...      1.536720       2.047399         1.421940   
566      -0.809117  ...      0.561361       1.374854         0.579001   
567       2.137194  ...      1.961239       2.237926         2.303601   
568      -0.820070  ...     -1.410893       0.764190        -1.432735   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
564    2.015301          0.378365          -0.273318         0.664512   
565    1.494959         -0.691230          -0.394820         0.236573   
566    0.427906         -0.809587           0.350735         0.326767   
567    1.653171          1.430427           3.904848         3.197605   
568   -1.075813         -1.859019          -1.207552        -1.305831   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
564              1.629151       -1.360158                -0.709091  
565              0.733827       -0.531855                -0.973978  
566              0.414069       -1.104549                -0.318409  
567              2.289985        1.919083                 2.219635  
568             -1.745063       -0.048138                -0.751207  

[5 rows x 31 columns]

Bu kod, "diagnosis" sütunundaki sınıfları (M: Malignant, B: Benign) sayısal değerlere dönüştürmek için LabelEncoder kullanır. Bu, sınıflandırma algoritması için giriş verisini hazırlar.

// m=1 b=0 olarak güncellendi


In [8]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
dataset["diagnosis"] = labelencoder.fit_transform(dataset["diagnosis"].values)


In [9]:
dataset.tail()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
564          1     2.110995      0.721473        2.060786   2.343856   
565          1     1.704854      2.085134        1.615931   1.723842   
566          1     0.702284      2.045574        0.672676   0.577953   
567          1     1.838341      2.336457        1.982524   1.735218   
568          0    -1.808401      1.221792       -1.814389  -1.347789   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
564         1.041842          0.219060        1.947285             2.320965   
565         0.102458         -0.017833        0.693043             1.263669   
566        -0.840484         -0.038680        0.046588             0.105777   
567         1.525767          3.272144        3.296944             2.658866   
568        -3.112085         -1.150752       -1.114873            -1.261820   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
564      -0.312589  ...      1.901185       0.117700         1.752563   
565      -0.217664  ...      1.536720       2.047399         1.421940   
566      -0.809117  ...      0.561361       1.374854         0.579001   
567       2.137194  ...      1.961239       2.237926         2.303601   
568      -0.820070  ...     -1.410893       0.764190        -1.432735   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
564    2.015301          0.378365          -0.273318         0.664512   
565    1.494959         -0.691230          -0.394820         0.236573   
566    0.427906         -0.809587           0.350735         0.326767   
567    1.653171          1.430427           3.904848         3.197605   
568   -1.075813         -1.859019          -1.207552        -1.305831   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
564              1.629151       -1.360158                -0.709091  
565              0.733827       -0.531855                -0.973978  
566              0.414069       -1.104549                -0.318409  
567              2.289985        1.919083                 2.219635  
568             -1.745063       -0.048138                -0.751207  

[5 rows x 31 columns]

In [10]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(dataset, test_size=0.3)


X_train = train.drop("diagnosis", axis=1): Bu satır, eğitim verilerinin bağımsız değişkenlerini (X_train) hazırlar. train veri çerçevesinden "diagnosis" sütununu çıkartarak (drop) bağımsız değişkenleri elde eder. "diagnosis" sütunu, veri setindeki hedef değişkeni temsil eder, bu nedenle bağımsız değişkenlerin bu sütun dışında olması gerekir.

y_train = train.loc[:, "diagnosis"]: Bu satır, eğitim verilerinin bağımlı değişkenini (y_train) hazırlar. train veri çerçevesindeki "diagnosis" sütununu (loc kullanarak) bağımlı değişken olarak belirler.

In [11]:
X_train = train.drop("diagnosis", axis=1)
y_train = train.loc[:, "diagnosis"]
X_test = test.drop("diagnosis", axis=1)
y_test = test.loc[:, "diagnosis"]


 scikit-learn kütüphanesinden Logistic Regression sınıflandırma modelini içe aktarır ve eğitim verileri üzerinde bu modeli eğitir.

In [13]:
from sklearn.linear_model import LogisticRegression
model_1 = LogisticRegression()
model_1.fit(X_train, y_train)


LogisticRegression()

In [14]:
predictions = model_1.predict(X_test)


In [15]:
from sklearn.metrics import confusion_matrix, classification_report
confusion_matrix(y_test, predictions)
print(classification_report(y_test, predictions))


              precision    recall  f1-score   support

           0       0.97      1.00      0.98       112
           1       1.00      0.93      0.96        59

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171



In [17]:
from sklearn.svm import LinearSVC
#sınıflandırma modelini içe aktarır

In [18]:
model_2 = LinearSVC()
model_2.fit(X_train, y_train)
predictions = model_2.predict(X_test)


In [20]:
from sklearn.metrics import confusion_matrix, classification_report
confusion_matrix(y_test, predictions)
print(classification_report(y_test, predictions))


              precision    recall  f1-score   support

           0       0.95      0.98      0.96       112
           1       0.96      0.90      0.93        59

    accuracy                           0.95       171
   macro avg       0.96      0.94      0.95       171
weighted avg       0.95      0.95      0.95       171



Lineer regresyon ve lojistik regresyon, iki farklı türde regresyon analizi yöntemidir ve temel olarak şu farklara sahiptirler:

Çıkış Değişkeni Türü:

Lineer Regresyon: Lineer regresyon, sürekli bir çıkış değişkenini tahmin etmek için kullanılır. Örneğin, ev fiyatlarını tahmin etmek için kullanılabilir. Çıkış değişkeni sayısal bir değerdir.
Lojistik Regresyon: Lojistik regresyon, iki kategoriden birine ait olma olasılığını tahmin etmek için kullanılır. Örneğin, bir hasta kanser taşıyıp taşımadığını tahmin etmek için kullanılabilir. Çıkış değişkeni genellikle 0 ile 1 arasında bir olasılık değeridir.
Model Tipi:

Lineer Regresyon: Lineer regresyon, bağımsız değişkenlerin doğrusal bir kombinasyonunu kullanarak çıkış değişkenini tahmin etmeye çalışır. İki veya daha fazla bağımsız değişken arasındaki ilişkiyi belirlemek için kullanılır.
Lojistik Regresyon: Lojistik regresyon, bağımsız değişkenlerin doğrusal bir kombinasyonunu sigmoid fonksiyonu içinde kullanarak çıkış değişkeni için bir olasılık tahmini yapar. Sınıflandırma problemleri için kullanılır.
Kullanım Alanları:

Lineer Regresyon: Lineer regresyon, regresyon problemlerinde kullanılır. Örneğin, gelirin yaş, eğitim seviyesi gibi faktörlere göre nasıl tahmin edilebileceğini anlamak için kullanılabilir.
Lojistik Regresyon: Lojistik regresyon, sınıflandırma problemlerinde kullanılır. Özellikle ikili sınıflandırma problemlerinde (örneğin, hasta/sağlıklı) kullanışlıdır.
Çıkış İfadesi:

Lineer Regresyon: Lineer regresyon, bir doğru denklemi şeklinde ifade edilir. Örneğin, "y = mx + b" gibi.
Lojistik Regresyon: Lojistik regresyon, log-odds (lojit) şeklinde ifade edilir. İkili sınıflandırma durumunda, "logit(p) = b0 + b1x1 + b2x2 + ... + bn*xn" gibi bir ifade kullanılır.
Hata Ölçüsü:

Lineer Regresyon: Lineer regresyonun hata ölçüsü genellikle MSE (Ortalama Kare Hatası) veya benzeri yöntemlerle hesaplanır.
Lojistik Regresyon: Lojistik regresyonun başarı ölçüsü genellikle doğruluk oranı, hassasiyet, duyarlılık gibi sınıflandırma performans metrikleri kullanılır.
Özetle, lineer regresyon sürekli bir çıkış değişkenini tahmin etmek için kullanılırken, lojistik regresyon sınıflandırma problemlerinde iki kategoriden birine ait olma olasılığını tahmin etmek için kullanılır. İki yöntem farklı problemler için tasarlanmıştır ve farklı matematiksel yapıya sahiptir.